### PR TITLE
Add egeniouss receiver.

### PR DESCRIPTION
**Pull request by mohsenD98**
# Enable Egeniouss Receiver Support

### Overview

This pull request implements support for Egeniouss receivers in the QField application. A new class, `EgenioussReceiver`, allows users to establish a TCP connection and interact with Egeniouss devices.

--- 

### New Class: EgenioussReceiver

The EgenioussReceiver class, derived from AbstractGnssReceiver, has been created. It facilitates a `TCP` connection to Egeniouss devices, using a fixed host address (`localhost`) and port (`1235`).

---

### Activation Instructions

The Egeniouss receiver is automatically enabled upon adding the Egeniouss plugin.

---

### Additional Information

Within the plugin, we have two key functions: `appWideEnabled()` and `appWideDisabled()`. These functions are invoked in QField when the plugin is enabled or disabled. They simply toggle the `egenioussEnabled` property as follows:

```C++
plugin.positioningSettings.egenioussEnabled = true;  // To enable
``` 
or
```C++
plugin.positioningSettings.egenioussEnabled = false; // To disable
```
By utilizing egenioussEnabled within the positioningSettings, we can control the visibility of the EgenioussReceiver.

---

### Download the Plugin
You can download the Egeniouss plugin [here](https://github.com/user-attachments/files/17098758/qfield-egeniouss-plugin.zip).

### CODE REVIEW

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for aac785171ee05d99e0190442a40bad26943e285f: [**arm64-android**](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield-dev-5637-48f49a24ee408601d91ee3f8468440d8932601f1-android-arm64.apk)
<details>
<summary>Other architectures</summary>

- [arm-neon-android](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield-dev-5637-48f49a24ee408601d91ee3f8468440d8932601f1-android-arm-neon.apk)
- [x64-android](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield-dev-5637-48f49a24ee408601d91ee3f8468440d8932601f1-android-x64.apk)
- [x86-android](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield-dev-5637-48f49a24ee408601d91ee3f8468440d8932601f1-android-x86.apk)
- [**all access arm64-android**](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield_all_access-dev-5637-48f49a24ee408601d91ee3f8468440d8932601f1-android-arm64.apk)
- [all access arm-neon-android](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield_all_access-dev-5637-48f49a24ee408601d91ee3f8468440d8932601f1-android-arm-neon.apk)
- [all access x64-android](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield_all_access-dev-5637-48f49a24ee408601d91ee3f8468440d8932601f1-android-x64.apk)
- [all access x86-android](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield_all_access-dev-5637-48f49a24ee408601d91ee3f8468440d8932601f1-android-x86.apk)
</details>

File: src/core/positioning/egenioussreceiver.h
Comment (nirvn) [inline src/core/positioning/egenioussreceiver.h:30]: You don't need that, it's in the abstractgnssreceiver.h header and you've made this class a friend.

File: src/core/positioning/egenioussreceiver.h
Comment (nirvn) [inline src/core/positioning/egenioussreceiver.h:32]: You don't need mReceivedData, let's just keep payload. Rename it to mPayLoad, and make it a QJsonObject (comment above)

File: src/core/positioning/egenioussreceiver.cpp
Comment (nirvn) [inline src/core/positioning/egenioussreceiver.cpp:9]:
```suggestion
  : AbstractGnssReceiver( parent ), 
    mAddress( address ), 
    mPort( port ), 
    mTcpSocket( new QTcpSocket( this ) )
```

File: src/core/positioning/egenioussreceiver.cpp
Comment (nirvn) [inline src/core/positioning/egenioussreceiver.cpp:17]: connect this in the constructor, we should connect/disconnect these every time.

File: src/core/positioning/egenioussreceiver.cpp
Comment (nirvn) [inline src/core/positioning/egenioussreceiver.cpp:26]: Don't wait for connection, listen to signals to change connected status.

File: src/core/positioning/egenioussreceiver.cpp
Comment (nirvn) [inline src/core/positioning/egenioussreceiver.cpp:46]: Don't wait, listen to signal

File: src/core/positioning/egenioussreceiver.cpp
Comment (nirvn) [inline src/core/positioning/egenioussreceiver.cpp:55]: Don't store the payload as a byte array, store is as a QJsonObject. You already create that object in onReadyRead, let's avoid the double-cost here.

File: src/core/positioning/egenioussreceiver.cpp
Comment (nirvn) [inline src/core/positioning/egenioussreceiver.cpp:57]: Then here you can just do if ( mPayload.isEmpty() ) { return detailsList; }

File: src/core/positioning/egenioussreceiver.cpp
Comment (nirvn) [inline src/core/positioning/egenioussreceiver.cpp:52]:
```suggestion
  QList<QPair<QString, QVariant>> detailsList;
```

File: src/core/positioning/egenioussreceiver.cpp
Comment (nirvn) [inline src/core/positioning/egenioussreceiver.cpp:106]: As the spec is developed, we'll need to make sure a lack of lat doesn't result is a 0.0 value but a quiet_NaN value.

File: src/core/positioning/positioning.cpp
Comment (nirvn) [inline src/core/positioning/positioning.cpp:197]: As discussed in chat, we don't want a port / address configuration for egeniouss. You'll have to move that back into tcp / udp conditional blocks.

File: src/core/positioning/egenioussreceiver.h
Comment (mohsenD98) [reply src/core/positioning/egenioussreceiver.h]: Correct :+1:

File: src/core/positioning/egenioussreceiver.h
Comment (mohsenD98) [reply src/core/positioning/egenioussreceiver.h]: Nice :+1:

File: src/core/positioning/egenioussreceiver.cpp
Comment (mohsenD98) [reply src/core/positioning/egenioussreceiver.cpp]: Correct :+1:

File: src/core/positioning/egenioussreceiver.cpp
Comment (mohsenD98) [reply src/core/positioning/egenioussreceiver.cpp]: Correct :+1:

File: src/core/positioning/egenioussreceiver.cpp
Comment (mohsenD98) [reply src/core/positioning/egenioussreceiver.cpp]: Correct :+1:

File: src/core/positioning/egenioussreceiver.cpp
Comment (mohsenD98) [reply src/core/positioning/egenioussreceiver.cpp]: Nice :+1:

File: src/core/positioning/egenioussreceiver.cpp
Comment (mohsenD98) [reply src/core/positioning/egenioussreceiver.cpp]: Yes :+1:

File: src/core/positioning/egenioussreceiver.cpp
Comment (mohsenD98) [reply src/core/positioning/egenioussreceiver.cpp]: how is this : 
`const double latitude = mPayload.value( "lat" ).toDouble() == 0 ? std::numeric_limits<double>::quiet_NaN() : mPayload.value( "lat" ).toDouble();`

File: src/core/positioning/egenioussreceiver.h
Comment (m-kuhn) [inline src/core/positioning/egenioussreceiver.h:1]: We normally do `#ifdef` include guards

File: src/qml/PluginManagerSettings.qml
Comment (nirvn) [inline src/qml/PluginManagerSettings.qml:18]: Hm, I think it'd be nicer if we implemented something slightly different and have the plugin manager call function _attached to the plugin Item {} itself_ when enabling/disabling a plugin. 

That way we a/ empower plugin authors, and b/ we avoid adding code in qgismobileapp.qml to handle a plugin-specific Uuid handling which I'm not fond of.

File: src/core/pluginmanager.cpp
Comment (nirvn) [inline src/core/pluginmanager.cpp:477]: methodIndex would be a better name. Or have methodExists be a bool and add >= 0 at the end of this line.

**Comment by nirvn** [comment]
@domi4484 , I'm good with this now. Do you want to have a look before we merge?

File: src/core/positioning/egenioussreceiver.cpp
Comment (domi4484) [inline src/core/positioning/egenioussreceiver.cpp:37]: It feels a bit strange to me that subclasses of postioning device have to manage `mSocketState` that way. mSocketState and mSocketStateString have a quite risk to get out of sync. Wouldn't be it a cleaner interface to reimplement `AbstractGnssReceiver::socketState()` and `socketStateString` in subclasses and get rid of the variables mSocketState and mSocketStateString?

File: src/core/positioning/egenioussreceiver.cpp
Comment (domi4484) [inline src/core/positioning/egenioussreceiver.cpp:42]: You could use `QVariantMap` to have dict like instead of plain pair list

File: src/core/positioning/egenioussreceiver.cpp
Comment (domi4484) [inline src/core/positioning/egenioussreceiver.cpp:109]: I guess with the current interface here you should set mSocketState and mSocketStateString as well

File: src/core/positioning/egenioussreceiver.cpp
Comment (domi4484) [inline src/core/positioning/egenioussreceiver.cpp:51]: For what stands `q`? Is a key in the json? Eventually to be put in a const to avoid having literals around

File: src/core/positioning/egenioussreceiver.cpp
Comment (domi4484) [inline src/core/positioning/egenioussreceiver.cpp:59]:
```suggestion
  if ( mReceivedData.size() < EgenioussReceiver::MINIMUM_DATA )
```
I would define consts for such kind of things

File: src/core/positioning/egenioussreceiver.cpp
Comment (domi4484) [inline src/core/positioning/egenioussreceiver.cpp:61]: What should we do here? maybe set mLastError? or just print some warnings?

File: src/core/positioning/egenioussreceiver.cpp
Comment (domi4484) [inline src/core/positioning/egenioussreceiver.cpp:65]: Here as well missing error handling

File: src/core/positioning/positioningdevicemodel.cpp
Comment (domi4484) [inline src/core/positioning/positioningdevicemodel.cpp:176]: I would put those prefixes into static const members. For `serial:`, `udp:`, and so on as well

File: src/core/positioning/tcpreceiver.cpp
Comment (domi4484) [inline src/core/positioning/tcpreceiver.cpp:122]: In this function also mSocketState and socket state string should be updated

File: src/qml/EgenioussDeviceChooser.qml
Comment (domi4484) [inline src/qml/EgenioussDeviceChooser.qml:1]: Would put this file into a egeniouss directory. Maybe into positioning/egeniouss and move as well other positioning related files in there to gain some structure

File: src/qml/PositioningSettings.qml
Comment (domi4484) [inline src/qml/PositioningSettings.qml:48]:
```suggestion
  property bool egenioussEnabled: false
```
But for `positioningActivated` we use Active instead of Enable. But in the code we see generally more often "Enabled". Sot not sure which one you should choose :smile:

**Review by domi4484 — COMMENTED** [review]
Great thank you @mohsenD98!
I left some comments, @nirvn  please take also a look maybe you don't agree on all of them :)

File: src/core/positioning/egenioussreceiver.cpp
Comment (mohsenD98) [reply src/core/positioning/egenioussreceiver.cpp:42]: `QVariantMap` does not maintain the order of its elements (i think it is implemented using a `Hash` table) but we need to maintain insertion order.

File: src/core/positioning/egenioussreceiver.cpp
Comment (mohsenD98) [reply src/core/positioning/egenioussreceiver.cpp:51]: Yes, it is in JSON and we assume it stands for `Quality(?!)` however in my tests it was always zero.    

if we are sure it is quality i can go for something like this: 

```C++
const QString KEY_QUALITY = "q";
detailsList.append(qMakePair(KEY_QUALITY, mPayload.value(KEY_QUALITY).toDouble()));
```

@nirvn this `KEY_QUALITY` or `KeyQuality` or `keyQuality` ?

File: src/core/positioning/egenioussreceiver.h
Comment (nirvn) [reply src/core/positioning/egenioussreceiver.h:1]: @domi4484 , let's maybe not do this right away. ATM we have our other device classes sitting in positioning (Bluetooth, TCP,  UDP, serial) here and having a directory for a single pair of .CPP, .H feels a bit too much. 

If we end up needing another class (eg egenioussutils) let's maybe consider moving all of it into an src/core/egeniouss directory)

File: src/core/positioning/egenioussreceiver.cpp
Comment (nirvn) [reply src/core/positioning/egenioussreceiver.cpp:37]: @domi4484 , not a bad idea. Can we do that in a follow up PR?

File: src/core/positioning/egenioussreceiver.cpp
Comment (nirvn) [reply src/core/positioning/egenioussreceiver.cpp:109]: Correct.

File: src/core/positioning/egenioussreceiver.cpp
Comment (nirvn) [reply src/core/positioning/egenioussreceiver.cpp:51]: At this moment let's leave it as is, that's merely to have an extra detail in. We'll polish this when the protocol progresses and we have actual values to handle.

File: src/core/positioning/egenioussreceiver.cpp
Comment (nirvn) [reply src/core/positioning/egenioussreceiver.cpp:59]: Good idea

File: src/core/positioning/positioningdevicemodel.cpp
Comment (nirvn) [reply src/core/positioning/positioningdevicemodel.cpp:176]: Good idea.

File: src/core/positioning/tcpreceiver.cpp
Comment (nirvn) [reply src/core/positioning/tcpreceiver.cpp:122]: Let's defer to different PR here to focus on egeniouss bits

File: src/qml/EgenioussDeviceChooser.qml
Comment (nirvn) [reply src/qml/EgenioussDeviceChooser.qml:1]: I suggest this structural change be kept in a separate PR

**Review by nirvn — APPROVED** [review]

**Pull request merged**
